pkg/controller/admissionchecks/preemption/preemption.go

```diff
@@ -0,0 +1,300 @@
+			c.trigger("event", "Workload with reservation was deleted", "wl", klog.KObj(oldWl))
+		}
+		return
+	}
+
+	// If quota gets released.
+	if workload.HasQuotaReservation(oldWl) && !workload.HasQuotaReservation(newWl) {
+		c.trigger("event", "Quota released", "wl", klog.KObj(oldWl))
+	}
+
+	// Is a preemption pending workload.
+	if apimeta.IsStatusConditionPresentAndEqual(newWl.Status.AdmissionChecks, constants.PreemptionAdmissionCheckName, metav1.ConditionUnknown) {
+		c.trigger("event", "Pending workload updated", "wl", klog.KObj(oldWl))
+	}
+}
+
+func (c *Controller) NotifyAdmissionCheckUpdate(oldAc, newAc *kueue.AdmissionCheck) {
+	// Adding or removing a check will lead to a workload update later on
+	// we should ignore those events here.
+	if oldAc != nil && newAc != nil && ptr.Deref(oldAc.Spec.PreemptionPolicy, "") != ptr.Deref(newAc.Spec.PreemptionPolicy, "") {
+		c.trigger("event", "Admission check preemption policy changed", "ac", klog.KObj(oldAc))
+	}
+}
+
+func (c *Controller) run(ctx context.Context) {
+	c.log.V(2).Info("Start run")
+
+	// If there is nothing to do at this point.
+	if !c.cache.ShouldCheckWorkloadsPreemption() {
+		// skip the snapshot creation and exit
+		return
+	}
+
+	snapshot := c.cache.Snapshot()
+	workloads := filterWorkloads(&snapshot)
+	for _, wl := range workloads {
+		//1. remove the workload from the snapshot
+		snapshot.RemoveWorkload(wl)
+		usage := totalRequestsForWorkload(wl)
+		needPreemption := resourcesNeedingPreemption(wl, usage, &snapshot)
+		log := c.log.WithValues("workload", klog.KObj(wl.Obj))
+		//2. check if preemption is sill needed
+		if len(needPreemption) == 0 {
+			// No more resources need to be freed to accommodate this workload.
+			// The preemption is successful.
+			if err := c.updateFnc(ctx, c.client, wl.Obj, true); err != nil {
+				log.V(2).Error(err, "Unable to update the check state to True")
+				c.trigger("retryUpdate", klog.KObj(wl.Obj))
+			} else {
+				log.V(2).Info("Preemption ended")
+			}
+		} else {
+			// Additional resources need to be freed.
+			targets := c.preemptor.GetTargetsForResources(wl, needPreemption, usage, &snapshot)
+			if len(targets) == 0 {
+				// There are no candidates for eviction, the preemption can no linger be done.
+				// The preemption failed.
+				if err := c.updateFnc(ctx, c.client, wl.Obj, false); err != nil {
+					log.V(2).Error(err, "Unable to update the check state to False")
+					c.trigger("retryUpdate", klog.KObj(wl.Obj))
+				} else {
+					log.V(2).Info("Preemption is no longer possible")
+				}
+			} else {
+				// Issue the evictions, the controller will run again once the evicted workloads are fully stopped.
+				count, err := c.preemptor.IssuePreemptions(ctx, targets, snapshot.ClusterQueues[wl.ClusterQueue])
+				if err != nil {
+					log.V(2).Error(err, "Unable to issue preemption")
+					c.trigger("retryEviction", klog.KObj(wl.Obj))
+				} else {
+					log.V(2).Info("Preemption triggered", "count", count)
```

Comment:
Why don't we need to update the workload with the check status?

Reply:
I see. So when is the check status updated? Do we expect the updates when triggered next time?

Reply:
yes, when the preemptee finishes execution,  we'll  check the fit an update the state accordingly.